Initial work to make hpc-intro configurable to different systems

aturner-epcc opened the pull request:
This is the first of a series of PRs to enable to longstanding HPC Carpentry goal of having the lessons configurable to different systems (hpc-intro already has this but the impact is more limited there).

This PR will be the largest as it involves shifting episode 12 away from the Lola Curious story back to the original format. I spoke to @psteinb before the Summer and he was OK with me doing this.

Throughout these changes I will ensure that the lesson is still available in the same form as it is now (i.e. the rendered lesson will look the same as it does now); with the exception of the major change to episode 12. 

This PR includes snippets that are not yet used by the template in later episodes - only episode 12 currently has the configurable parts at the moment but I hope to update in the others over the next few days via further PRs. The currently unused snippets are for the EPCC Cirrus system but I will be creating snippets to make sure the rendered lesson does not change as I go on.

Finally, I will add instructions to the README on how to configure the lesson for different systems.

aturner-epcc commented:
@CarrieBrown I have updated the PR to create the snippets library as described and also added initial notes on how to customise into _config.yml. Plan to update the other episodes to make them configurable over the next couple of days.

I can keep everything in this PR if you want rather than merging multiple times.

tkphd commented:
@aturner-epcc, please never commit directly to the `gh-pages` branch. Conduct your work on a new branch (`git checkout -b experimental_branch_name`, or on GitHub, click Branch: gh-pages, type in the experimental branch name you'd like, and click the Create Branch tooltip), and push that to your personal fork. This will prevent your fork from diverging from the upstream repository, and produce a lot less headache for future you :smile:

aturner-epcc commented:
@tkphd OK, I can cancel this PR and issue a new one from a branch. Do I issue the PR from my branch to gh-pages in the main repo or get my new branch in the main repo then issue a PR from the branch in the main repo to gh-pages in the main repo?

tkphd commented on _extras/snippets_library/EPCC_Cirrus_pbs/14/depend_demo.snip:
````diff
@@ -0,0 +1,47 @@
+```
+{: .output}
+
+So using `module unload` "un-loads" a module along with it's dependencies.
````
"it's" -> "its"

aturner-epcc replied:
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

tkphd commented on _extras/snippets_library/EPCC_Cirrus_pbs/_config_options.yml:
```diff
@@ -0,0 +1,19 @@
+sched_submitopt: "-A tc001 -q R387726"
+sched_stat: "qstat"
+sched_uopt: "-u yourUsername"
+sched_del: "qdel"
```
`sched_statu` is undefined.

aturner-epcc replied:
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

tkphd commented:
@aturner-epcc, you'll issue the PR against the main `gh-pages` branch in this repo. Once it's approved, the PR will get merged into `gh-pages`, as it should be; then everyone who pulls from this repo into their local fork will have a clean update. 

You don't necessarily have to close this PR and start over: the "clean" alternative is to make changes on your fork's `gh-pages` branch until this gets merged, then delete your fork since it's diverged, then fork this repo over again. It's not the preferred workflow, and I don't recommend it moving forward, but it'll work this time.

aturner-epcc commented:
@tkphd Thanks for the explanation. The original repo is an upstream of my fork so I think I can manage the fetch and merge once my changes are in without deleting my fork. 

Thanks also for taking the time to look through and producing the detailed comments. I will get to work on fixing them!

tkphd commented:
Happy to help!

aturner-epcc commented on _config.yml:
```diff
@@ -61,6 +61,37 @@ instructor_pre_survey: "https://www.surveymonkey.com/r/instructor_training_pre_s
+sched_submit: "sbatch"
+sched_submitopt: ""
+sched_stat: "squeue"
+sched_uopt: "-u yourUsername"
```
Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

aturner-epcc commented on _config.yml:
```diff
@@ -61,6 +61,37 @@ instructor_pre_survey: "https://www.surveymonkey.com/r/instructor_training_pre_s
+#     _extras/snippet_library/EPCC_Cirrus_pbs/_config_options.yml
+#   and:
+#     cp -r _extras/snippet_library/EPCC_Cirrus_pbs/* _includes/snippets
+#
```
Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -1,71 +1,49 @@
 ---
 
-## The Story
+## What is a an HPC system?
```
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -1,71 +1,49 @@
+too big for any one computer.
 
-## Going remote
+![The cloud is made of Linux](../fig/linux-cloud.jpg)
```
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

aturner-epcc commented on _config.yml:
```diff
@@ -61,6 +61,37 @@ instructor_pre_survey: "https://www.surveymonkey.com/r/instructor_training_pre_s
+sched_info: "sinfo"
+sched_comment: "#SBATCH"
+sched_nameopt: "-J"
+sched_hist: "sacct  -u yourUsername"
```
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

aturner-epcc commented on _episodes/12-cluster.md:
````diff
@@ -82,65 +60,99 @@ permissions for on the cluster.
+> Now we'll compare the size of your computer with the size of the head node: To see the number of
+> processors, run:
+>
+> ```
````
@tkphd Fixed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/579a4bdd87c290832744133096f211a6555b24bb

tkphd replied:
Better, but it looks like these are supposed to run on the student's local machine, not the login node. Can we use a local prompt variable, or hard-code something akin to the following?
```yaml
local_prompt: "[user@laptop ~]$ "
```

aturner-epcc commented:
@tkphd I think I have fixed all the problems you raised. Let me know if I missed anything.

I think that is this initial work complete now so the lesson should now be able to be customised for different systems using the config options and snippets. There is still some work that could be done to make it more easily portable but I will raise these as separate issues as they do not stop it being used as it is.

tkphd commented on _includes/snippets/12/explore.snip:
````diff
@@ -0,0 +1,10 @@
+> run. Try running this command to see the name, CPUs and memory available on the worker nodes:
+>
+> ```
+> sinfo -n aci-377 -o "%n %c %m"
````
Missing a prompt

tkphd commented:
This may be a local config problem, but I cannot serve the site locally:

```bash
$ jekyll serve -t
Configuration file: hpc-intro/_config.yml
Configuration file: hpc-intro/_config.yml
            Source: hpc-intro
       Destination: hpc-intro/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
                    done in 1.0 seconds.
 Auto-regeneration: enabled for 'hpc-intro'
Configuration file: hpc-intro/_config.yml
/usr/lib/ruby/2.1.0/socket.rb:232:in `getaddrinfo': getaddrinfo: Name or service not known (SocketError)
        from /usr/lib/ruby/2.1.0/socket.rb:232:in `foreach'
        from /usr/lib/ruby/2.1.0/socket.rb:459:in `tcp_server_sockets'
        from /usr/lib/ruby/2.1.0/webrick/utils.rb:75:in `create_listeners'
        from /usr/lib/ruby/2.1.0/webrick/server.rb:132:in `listen'
        from /usr/lib/ruby/2.1.0/webrick/server.rb:113:in `initialize'
        from /usr/lib/ruby/2.1.0/webrick/httpserver.rb:45:in `initialize'
        from /var/lib/gems/2.1.0/gems/jekyll-3.3.1/lib/jekyll/commands/serve.rb:105:in `new'
        from /var/lib/gems/2.1.0/gems/jekyll-3.3.1/lib/jekyll/commands/serve.rb:105:in `start_up_webrick'
        from /var/lib/gems/2.1.0/gems/jekyll-3.3.1/lib/jekyll/commands/serve.rb:51:in `process'
        from /var/lib/gems/2.1.0/gems/jekyll-3.3.1/lib/jekyll/commands/serve.rb:39:in `block (2 levels) in init_with_program'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `call'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
        from /var/lib/gems/2.1.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
        from /var/lib/gems/2.1.0/gems/jekyll-3.3.1/exe/jekyll:13:in `<top (required)>'
        from /usr/local/bin/jekyll:23:in `load'
        from /usr/local/bin/jekyll:23:in `<main>'
```

Serving the trunk (`gh-pages`) works, *i.e.*,

```bash
$ jekyll serve
Configuration file: hpc-intro/_config.yml
Configuration file: hpc-intro/_config.yml
            Source: hpc-intro
       Destination: hpc-intro/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
                    done in 1.011 seconds.
 Auto-regeneration: enabled for 'hpc-intro'
Configuration file: hpc-intro/_config.yml
    Server address: http://127.0.0.1:4000/
  Server running... press ctrl-c to stop.
```

aturner-epcc commented:
@tkphd Fixed up the issues with missing prompts in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/a2395cd04178ab7c1a15c298504a24d92b930895

`jekyll serve` issue is because I am a doofus! I updated the name of the HPC system to be set by the config variable `host` but, of course, this is a config option used by Jekyll. I fixed by changing this to `host_name`.

aturner-epcc commented:
Thanks @tkphd 

@psteinb I will let you do your review before I merge this

psteinb commented:
Thanks, I'll sit down over it.

psteinb commented on _episodes/12-cluster.md:
```diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+All of a HPC system's nodes have the same components as your own laptop or desktop: *CPUs* (sometimes
+also called *processors* or *cores*), *memory* (or *RAM*), and *disk* space. CPUs are a computer's
+tool for actually running programs and calculations. Information about a current task is stored in
+the computer's memory. Disk is a computer's long-term storage for information it will need later.
```
Maybe reword: `Disk is a computer's long-term storage for information it will need later.`
A suggestion: `Disk refers to all storage that can be accessed like a filesystem. This is generally storage that can hold data permanently, i.e. data is still there even if the computer has been restarted.`

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

psteinb commented on _episodes/12-cluster.md:
```diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
 
-The admin continues to encourage Lola to look around. She explains that all of a 
-cluster's nodes have similar components as Lola's laptop or workstation. 
+All of a HPC system's nodes have the same components as your own laptop or desktop: *CPUs* (sometimes
```
This paragraph is super essential. It might need some figures or pictures that people can (re)use during the course.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

psteinb commented on _episodes/12-cluster.md:
````diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+> processors, run:
+>
+> ```
+> {{ site.host_prompt}} nproc --all
````
Can we be sure that `nproc` is installed everywhere?

aturner-epcc replied:
I think the only honest answer to that is "no". However, it is available on all the systems I have tried it on (even ARCHER, which is an ancient Cray XC30 running SLES 11!).

psteinb commented on _episodes/12-cluster.md:
````diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+> or
+>
+> ```
+> {{ site.host_prompt}} lscpu
````
This will confuse people at this point. I'd suggest to remove the whole `lscpu` part.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

psteinb commented on _episodes/12-cluster.md:
````diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+>
+> or for more details: 
+>
+> ```
````
`cat /proc/meminfo` confuses people more than necessary. I suggest to put it together with `cat /proc/cpuinfo` (or `lscpu`) these into a callout box, so that there is a visual hint that this information is not required in order to proceed.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/commit/8e2ca7059afd9b0080cab714be51e85c712790cf

psteinb commented on _episodes/12-cluster.md:
```diff
@@ -158,242 +168,11 @@ $ free -g
-
+With all of this in mind, we will now cover how to talk to the cluster's scheduler, and use it to
+start running our scripts and programs!
 
```
There are exercises missing here. One idea: 
> Compare your laptop's number of processors and memory with the numbers you see on the login node! Share your findings with the others.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

psteinb commented on _episodes/15-transferring-files.md:
```diff
@@ -138,6 +86,23 @@ Hit "Quickconnect" to connect! You should see your remote files appear on the ri
 screen. You can drag-and-drop files between the left (local) and right (remote) sides of the screen
 to transfer files.
 
+## Compressing files
```
What is the reason for this section here? I am wondering, if this content about compressing files is required or essential for the introduction to HPC.

aturner-epcc replied:
My justification would be that archiving (often involving compression) is a key part of being able to transfer data to/from remote systems. It is really important to build in this good habit for new users as a lot of problems researchers have with HPC are manipulating data between different places in their research workflow. However, re-reading this section I think the heading "Compression" is the wrong one as it is the "Archiving" of multiple files into a single file is *the* key concept we want to get across here. 

I will have a go at rewording/refocussing.

aturner-epcc replied:
I have updated this in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/e0abb4c06b6e0fae4cccba1d3211b8bf9d06f830

Take a look and let me know if you think this is better.

psteinb replied:
Super, if I could buy you a coffee, I'd do so. e0abb4c turns the tide completely and gives it quite some relevance. thanks for putting time into this.

psteinb commented on _episodes/17-responsiblity.md:
```diff
@@ -1,74 +1,243 @@
-  {% for queue in site.workshop_interactive_queues %}{{ queue }}, {% endfor %} etc.  
-  This is generally a better way to run small tests or compile programs.
+Don’t run jobs on the login node (though quick tests are generally fine). A “quick test” is
+generally anything that uses less than 4GB of memory, 4 CPUs, and 10 minutes of time. If you
```
`4GB of memory, 4 CPUs, and 10 minutes of time` from my experience, beginners don't know how to measure memory consumption, number of CPUs used may be doable as well as runtime of a process. Therefor it's super tricky to make these kind of recommendations. I'd personally suggest to concentrate on `10 minutes of time` only here. But I am open for suggestions.

aturner-epcc replied:
I agree about the memory point and think you are right: using many cores for 10s will probably not be noticed by other users as it is so transient. I will rephrase and probably reduce to 5 minutes of time.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/152f118c823ddf914b2c9b09035af1a649f8cc4d

psteinb commented on _episodes/17-responsiblity.md:
```diff
@@ -1,74 +1,243 @@
+
+On very busy resources you may wait many days in a queue for your job to fail within 10 seconds
+of starting due to a trivial typo in the job script. This is extremely frustrating! Most
+systems provide small, short queues for testing that have short wait times to help you 
```
the concept of queues may not exist on the cluster at hand. Maybe just say `Most systems provide smaller resources for testing`

aturner-epcc replied:
Addressed https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/152f118c823ddf914b2c9b09035af1a649f8cc4d

psteinb commented on _episodes/17-responsiblity.md:
```diff
@@ -1,74 +1,243 @@
+is strongly recommended to pack the files into a larger *archive* file for long term storage and
+transfer. A single large file makes more efficient use of the file system and is easier to move,
+copy and transfer because significantly fewer meta-data operations are required. Archive files can
+be created using tools like `tar`, `cpio` and `zip`. We are going to look at `tar` as it is the 
```
don't confuse people with completeness, please remove `cpio` if it is not needed later.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/152f118c823ddf914b2c9b09035af1a649f8cc4d

psteinb commented on _episodes/17-responsiblity.md:
```diff
@@ -1,74 +1,243 @@
+be created using tools like `tar`, `cpio` and `zip`. We are going to look at `tar` as it is the 
+most commonly used.
+
+The `tar` command packs files into a "Tape ARchive" format intended for backup purposes. To create
```
while the reference to `Tape ARchive` is valid, it can confuse users that really have access to a tape archive from within their HPC cluster. A link to wikipedia is enough.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/152f118c823ddf914b2c9b09035af1a649f8cc4d

It is not ideal that we introduce tar in the transfer episode and then talk about it again here but I think this can be addressed in a future update. Let me know if you agree or disagree with this approach. If you agree, I will raise an Issue on this.

psteinb replied:
fine, please create the issue so that we don't forget.

psteinb commented on _extras/18-nextsteps.md:
```diff
@@ -0,0 +1,117 @@
+> questions (both technical and non-technical).
+{: .challenge}
+
+> ## Exploring the performance of GROMACS
```
This is a very life-science based exercise and it is very involved. I think this is worth a capstone project. I would call it as such. Moreover, for people that have no professional relationship with Gromacs, the outcome of the exercise might be irrelevant and thus demotivating.

aturner-epcc replied:
This is why it is in `_extras` and not part of the main lesson. This was a useful resource that I created for a particular delivery of HPC Carpentry and thought it might be useful more widely but did not fit within the main lesson.

If you think the inclusion in `_extras` is still too strong I am happy to take it out.

I could bring it back as its own single lesson ('hpc-bootstrap'?) that could be appended to other lessons as needed.

I would be interested in your opinion here.

psteinb commented on _extras/18-nextsteps.md:
```diff
@@ -0,0 +1,117 @@
+> Please ask for more information on these options from a helper!
+{: .challenge}
+
+> ## Running many serial BLAST+ analyses in parallel
```
This is a very life-science based exercise and it is very involved. I think this is worth a capstone project. I would call it as such. Moreover, for people that have no professional relationship with Blast+, the outcome of the exercise might be irrelevant and thus demotivating. If the value of the outcome would be stressed more, I'll say I like the conjunction with gnu-parallel.

psteinb replied:
Ok, I didn't understand it that way. if it is in extras and introduced as such, I am super happy to make it part of this repo.

psteinb commented on _extras/snippets_library/EPCC_Cirrus_pbs/12/explore.snip:
````diff
@@ -0,0 +1,10 @@
+> run. Try running this command to see the name, CPUs and memory available on the worker nodes:
+>
+> ```
+> {{ site.host_prompt }} pbsnodes r1i0n32
````
Add a comment that the node name is provided by the instructor.

aturner-epcc replied:
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/d61b43ba0fab708f940d9e87801e074ad95432db

aturner-epcc commented:
@psteinb Just to clarify, does your last comment mean I should go ahead and merge or wait for you to review?

psteinb commented:
Just finished my review. Please have a look at the comments. After that, you can merge.

tkphd commented:
Excellent comments, @psteinb.

psteinb commented:
@aturner-epcc @tkphd how should we go about the review? Some people are waiting on this PR being merged. I am happy to help if needed.

aturner-epcc commented:
@psteinb Sorry, been swamped with stuff over the last week. I think I have your points covered, just need to put together the response. Should have it by the end of this week - if I do not hit the end of the week then I will ask for help

psteinb commented:
fair enough. go for it and thanks for your hard work.
P

aturner-epcc commented on _config.yml:
```diff
@@ -61,6 +61,38 @@ instructor_pre_survey: "https://www.surveymonkey.com/r/instructor_training_pre_s
+local_prompt: "[user@laptop ~]$"
+sched_name: "SLURM"
+sched_submit: "sbatch"
+sched_submit_options: ""
```
You can make this work with PBS, see the PBS example for Cirrus at EPCC in `_extras/snippets_library/EPCC_Cirrus_pbs`. The values are:

```
sched_interactive: "qsub"
sched_flag_interactive: "-IVl select=1:ncpus=1"
```

and

```
sched_hist: "qstat -x"
sched_flag_histdetail: "-f"
```

and the corresponding snippets in the appropriate subdirectories. When I fist designed this it was *exactly* to make it work with PBS (which is what we use locally) and I have successfully taught hpc-intro with the PBS substitutes, see:

https://archer-cse.github.io/2018-12-06-turing-hpcintro/

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -1,71 +1,47 @@
-- "Running a process on the cluster requires using the scheduler."
+- "An HPC system is a set of networked machines."
+- "HPC systems typically provides login nodes and a set of worker nodes."
+- "The resources available on different nodes can be different."
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -1,71 +1,47 @@
-Laptop at home for her studies, so the idea of using a super computer appears a bit intimidating to
-her. Her supervisor notices her anxiety and tells her that she will receive an introduction to the
-super computer after she has requested an account on the cluster. 
+The words "cloud", "cluster", and "high-performance computing" get thrown around a lot. So what do
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -1,71 +1,47 @@
-explains that Lola will use a small to mid-range HPC cluster.
+The *cloud* is a generic term commonly used to refer to remote computing resources of any kind --
+that is, any computers that you use but are not right in front of you. Cloud can refer to
+webservers, remote storage, API endpoints, as well as more traditional "compute" resources. An
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+entire computing cluster. So what's really happening? What computer have we logged on to? The name
+of the current computer we are logged onto can be checked with the `hostname` command. (You may also
+notice that the current hostname is also part of our prompt!)
+
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
+quick tests. It should never be used for doing actual work.
+
+The real work on a cluster gets done by the *worker* (or *compute*) *nodes*. Worker nodes come in
+many shapes and sizes, but generally are dedicated to doing all of the heavy lifting that needs
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented on _episodes/12-cluster.md:
```diff
@@ -82,65 +58,99 @@ permissions for on the cluster.
-$ logout
-~~~
-{: .language-bash}
+**This is an important point to remember: files saved on one node (computer) are often available
```
Addressed in https://github.com/hpc-carpentry/hpc-intro/pull/101/commits/8e2ca7059afd9b0080cab714be51e85c712790cf

aturner-epcc commented:
@psteinb thanks for taking the time to look through and make such constructive comments, your work is much appreciated. I think have addressed most of the comments and added explanations to a couple of points. Can you take a look through and let me know what you think?
